File: Day 005/Problem61.RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    public:
        ListNode* rotateRight(ListNode* head, int k) {
            if (!head || !head->next) return head;
            
            int listSize = 0;
            ListNode* temp = head;
            while (temp) {
                listSize++;
                temp = temp->next;
            }
    
            k = k % listSize;
    
            for (int i = 0; i < k; i++) {
                int prevVal = head->val;
    
                ListNode* ptr = head;
                while (ptr && ptr->next) {
                    int nextVal = ptr->next->val;
                    ptr->next->val = prevVal;
                    prevVal = nextVal;
                    ptr = ptr->next;
                }
    
                head->val = prevVal;
            }
            return head;
        }
    };

File: Day 090/Problem 61. Rotate List.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}

class Solution {
    private ListNode rotate(ListNode head) {
        ListNode ptr = head;
        while (ptr.next.next != null) {
            ptr = ptr.next;
        }

        ListNode tail = ptr.next;
        ptr.next = null;

        tail.next = head;
        return tail;
    }

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) return head;

        int size = 0;
        for (ListNode ptr = head; ptr != null; ptr = ptr.next) {
            size++;
        }

        k = k % size;

        for (int i = 0; i < k; i++) {
            head = rotate(head);
        }

        return head;
    }
}